src/promptflow/promptflow/_core/tools_manager.py

```diff
@@ -53,6 +54,23 @@ def collect_tools_from_directory(base_dir) -> dict:
     return tools
 
 
+def try_to_set_input_index_for_tool(tool):
+    tool_type = tool.get("type")
+    if _should_add_index(tool_type) and "inputs" in tool:
+        inputs_dict = tool["inputs"]
```

Comment:
The inputs is a dict, how can we make sure the dict respect the order user defined in yaml?

Reply:
The insertion-order preservation nature of dict objects has been declared to be an official part of the Python language spec after python 3.7 ([doc link](https://docs.python.org/3/whatsnew/3.7.html))
![image](https://github.com/microsoft/promptflow/assets/95729303/04f91154-9935-49d6-ab9e-9d5b8d07dd5a)
Additionally, the order was also maintained in all the tested cases.

Reply:
![image](https://github.com/microsoft/promptflow/assets/95729303/d7117ae1-2203-4a68-8692-c31c13182e71)
It uses the ruamel.yaml to load the yaml file, checked [the doc](https://yaml.readthedocs.io/en/latest/overview/#overview), it can keep the key order 
![image](https://github.com/microsoft/promptflow/assets/95729303/c3616878-32f8-4f27-b8c4-11256b6fc45f)